network.py

```diff
@@ -28,37 +34,38 @@ def add_network_layer(self, layer: Layer) -> None:
 
     def train(self, training_set: pd.DataFrame, visualize: bool = False) -> None:
         current_iteration = 0
-        while not self.__stop_condition_met(current_iteration):
+        mean_error = self.__error_threshold
+        while not self.__stop_condition_met(current_iteration, mean_error):
             current_iteration += 1
-            sum_error = 0
+            if current_iteration == self.__epoch_no:
+                pass
             training_set = training_set.sample(frac=1)  # shuffle training set
             x_train, y_train = training_set.values[:, :-1], training_set.values[:, -1]
+            mean_error = 0.0
             input_value: np.ndarray
             expected_result: float
+            sample_len = len(x_train)
             for (input_value, expected_result) in zip(x_train, y_train):
                 outputs = self.__forward_pass(input_value)
-                expected = [0 for i in range(len(outputs))]
+                expected = [0 for _ in range(len(outputs))]
                 expected[int(expected_result) - 1] = 1
-                sum_error += sum([(expected[i] - outputs[i]) ** 2 for i in range(len(expected))]) # TODO: This error probably doesn't correspond with calculated one?
+                mean_error += self.__error_function(outputs[int(expected_result) - 1], 1) / sample_len
```

Comment:
Maybe divide by `sample_len` after the loop? Or it leads to overflow or other issues?

Reply:
Will do but in other branch to avoid conflicts